### PR TITLE
twoliter: external kit test project and include rpm repositories

### LINKED ISSUE #257: buildsys: use external kits
Buildsys/Dockerfile/rpm2img/rpm2kit are affected. They need to know about the location of external kit repos and make these available in the repo creation and dnf install stages.

Edit:

Here is Ben's diff demonstrating how this could be done:

```diff
diff --git a/twoliter/embedded/build.Dockerfile b/twoliter/embedded/build.Dockerfile
index a27b728..2c75103 100644
--- a/twoliter/embedded/build.Dockerfile
+++ b/twoliter/embedded/build.Dockerfile
@@ -146,12 +146,17 @@ RUN --mount=target=/host \
     for kit in ${KIT_DEPENDENCIES} ; do \
       KIT_REPOS+=("--repofrompath=${kit},/host/build/kits/${kit}/${ARCH}" --enablerepo "${kit}") ; \
     done && \
+    declare -a EXTERNAL_KIT_REPOS && \
     echo "${KIT_REPOS[@]}" && \
+    for kit in ${EXTERNAL_KIT_DEPENDENCIES} ; do \
+      EXTERNAL_KIT_REPOS+=("--repofrompath=${kit},/host/build/external-kits/${kit}/${ARCH}" --enablerepo "${kit}") ; \
+    done && \
     dnf -y \
         --disablerepo '*' \
         --repofrompath repo,./rpmbuild/RPMS \
         --enablerepo 'repo' \
         "${KIT_REPOS[@]}" \
+        "${EXTERNAL_KIT_REPOS[@]}" \
         --nogpgcheck \
         --forcearch "${ARCH}" \
         builddep rpmbuild/SPECS/${PACKAGE}.spec
```

### PR DESCRIPTION
**Pull request by jmt-lab**
**Issue number:** 257

Closes #257

**Description of changes:**

* Adds a test project for external kits
* Adds discovery of external kit repositories

**Testing done:**

* Built local kit project and manually published the core-kit oci image to a private ecr
* Changed external-kit test project Twoliter.toml to use the private ecr for custom-vendor
* twoliter update
* twoliter fetch --arch=aarch64
* twoliter build kit extra-1-kit --arch=aarch64

extra-1-kit built successfully with pkg-b dependening on pkg-a from external core-kit

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: tests/projects/external-kit/.dockerignore [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/.git
+/.gomodcache
+/build/*
+!/build/rpms/
+/build/rpms/*
+!/build/rpms/*.rpm
```
Comment (webern) [inline tests/projects/external-kit/.dockerignore:6]: You can delete this file because Twoliter is bringing its own .dockerignore now (so this one isn't being used).

File: tests/projects/external-kit/README.md [hunk 2]
```diff
@@ -0,0 +1 @@
+This project represents a project structure with a single local kit with no external kit dependencies.
```
Comment (webern) [inline tests/projects/external-kit/README.md]: Update this to describe the purpose of the `external-kit` test project.

File: tests/projects/external-kit/Twoliter.toml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+registry = "public.ecr.aws/bottlerocket"
+
+[vendor.custom-vendor]
+# We need to figure out how we can do integration testing with this
+registry = "public.ecr.aws/bottlerocket"
+
+[[kit]]
+name = "core-kit"
+version = "0.1.0"
+vendor = "custom-vendor"
```
Comment (webern) [inline tests/projects/external-kit/Twoliter.toml]:
```suggestion
vendor = "custom-vendor"

```

File: tests/projects/external-kit/packages/pkg-b/pkg-b.spec [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+Name: %{_cross_os}pkg-b
+Version: 0.0
+Release: 0%{?dist}
+Summary: pkg-b
+License: Apache-2.0 OR MIT
+URL: https://github.com/bottlerocket-os/bottlerocket
+
+Source100: pkg-b.txt
+
+BuildRequires: %{_cross_os}pkg-a
```
Comment (webern) [inline tests/projects/external-kit/packages/pkg-b/pkg-b.spec:14]:
```suggestion
# This package comes from the core-kit defined in the local-kit project
BuildRequires: %{_cross_os}pkg-a
```

File: twoliter/embedded/Makefile.toml [hunk 5]
```diff
@@ -1642,6 +1642,7 @@ script_runner = "bash"
 script = [
 '''
 rm -rf ${BUILDSYS_KITS_DIR}
+rm -rf ${BUILDSYS_EXTERNAL_KITS_DIR}
```
Comment (webern) [inline twoliter/embedded/Makefile.toml]: Also create the directory and define the variable (see my PR)

File: tests/projects/external-kit/Twoliter.toml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+name = "bottlerocket-sdk"
+vendor = "bottlerocket"
+version = "0.41.0"
+
+[vendor.bottlerocket]
+registry = "public.ecr.aws/bottlerocket"
+
+[vendor.custom-vendor]
+# We need to figure out how we can do integration testing with this
+registry = "public.ecr.aws/bottlerocket"
```
Comment (webern) [inline tests/projects/external-kit/Twoliter.toml:14]: This is fine for now, but we will want to have some way to run external-kit tests and we won't be pushing these artifacts to our production ECR domain.

File: twoliter/embedded/build.Dockerfile [hunk 7]
```diff
@@ -138,11 +138,20 @@ RUN --mount=target=/host \
       KIT_REPOS+=("--repofrompath=${kit},/host/build/kits/${kit}/${ARCH}" --enablerepo "${kit}") ; \
     done && \
     echo "${KIT_REPOS[@]}" && \
+    declare -a EXTERNAL_KIT_REPOS && \
+    EXTERNAL_KITS=$(jq -c '.kit[]' < /host/build/external-kits/external-kit-metadata.json) && \
+    for kit in ${EXTERNAL_KITS} ; do \
+       KIT_NAME=$(jq '.vendor + "-" + .name' <<< "${kit}" | tr -d '"') && \
```
Comment (cbgbt) [inline twoliter/embedded/build.Dockerfile]: I believe you don't need this `tr` if you use `-r` (`--raw-output`) with `jq`.

File: twoliter/embedded/build.Dockerfile [hunk 8]
```diff
@@ -138,11 +138,20 @@ RUN --mount=target=/host \
...
+       REPO_PATH=$(jq "\"/host/build/external-kits/\" + .vendor + \"/\" + .name + \"/${ARCH}\"" <<< "${kit}" | tr -d '"') && \
+       EXTERNAL_KIT_REPOS+=("--repofrompath=${KIT_NAME},${REPO_PATH}" --enablerepo "${KIT_NAME}"); \
+    done && \
+    echo "${EXTERNAL_KIT_REPOS[@]}" && \
```
Comment (cbgbt) [inline twoliter/embedded/build.Dockerfile:162]: Is this just for debugging?

File: tests/projects/external-kit/Twoliter.toml (hunk 3 above)
Comment (bcressey) [inline tests/projects/external-kit/Twoliter.toml]: configure your editor to end files with a newline please!

File: twoliter/embedded/build.Dockerfile (hunk 7 above)
Comment (bcressey) [inline twoliter/embedded/build.Dockerfile]: The JSON parsing is making this too complicated. Pass the list of external kits in a space separated variable like `KIT_DEPENDENCIES` please.

**Review by bcressey — CHANGES_REQUESTED** [review]

File: twoliter/embedded/build.Dockerfile (hunk 8 above)
Comment (jmt-lab) [reply twoliter/embedded/build.Dockerfile:162]: Just following the pattern for previous block

**Comment by jmt-lab** [comment]
* Addressed Comments
* Rebased
* Fixed a few bugs here and there when testing
* Tested build kit again

File: twoliter/embedded/build.Dockerfile [hunk 9]
```diff
@@ -152,11 +153,19 @@ RUN --mount=target=/host \
       KIT_REPOS+=("--repofrompath=${kit},/host/build/kits/${kit}/${ARCH}" --enablerepo "${kit}") ; \
     done && \
     echo "${KIT_REPOS[@]}" && \
+    declare -a EXTERNAL_KIT_REPOS && \
+    for kit in ${EXTERNAL_KIT_DEPENDENCIES} ; do \
+       REPO_NAME="$(tr -s '/' '-' <<< "${kit}")" && \
```
Comment (bcressey) [inline twoliter/embedded/build.Dockerfile:158]: Is this transform needed? Seems like the variable can list kits in the canonical `vendor/kit` format.

File: tools/buildsys/src/manifest.rs [hunk 10]
```diff
@@ -322,11 +324,44 @@ impl Manifest {
         Ok(kits)
     }
 
+    /// List all external kits needed for the build by vendor-kit
```
Comment (bcressey) [inline tools/buildsys/src/manifest.rs]: nit: not sure what "by vendor-kit" means

File: twoliter/src/lock.rs [hunk 11]
```diff
@@ -222,13 +222,14 @@ impl OCIArchive {
...
-        if digest == self.digest {
-            return Ok(());
+        if digest_file.exists() {
+            let digest = read_to_string(&digest_file).await.context(format!(
+                "failed to read digest file at {}",
+                digest_file.display()
+            ))?;
+            if digest == self.digest {
+                return Ok(());
+            }
```
Comment (bcressey) [inline twoliter/src/lock.rs:232]: These fixes should be in a separate `twoliter` commit.

File: tests/projects/external-kit/.dockerignore [hunk 12]
```diff
@@ -0,0 +1,10 @@
+/.git
+/.gomodcache
+/build/*
+!/build/rpms/
+/build/rpms/*
```
Comment (bcressey) [inline tests/projects/external-kit/.dockerignore:5]: Is this file going away? The previous comment about it is resolved.

**Review by bcressey — APPROVED** [review]
No blocking feedback. Nice!

File: twoliter/embedded/build.Dockerfile (hunk 9 above)
Comment (jmt-lab) [reply twoliter/embedded/build.Dockerfile:158]: The transformation is for the repository name in dnf/yum, I wasn't sure / are allowed for repository names

File: tools/buildsys-config/src/lib.rs [hunk 13]
```diff
@@ -6,6 +6,7 @@ pub const EXTERNAL_KIT_DIRECTORY: &str = "build/external-kits";
 pub const EXTERNAL_KIT_METADATA: &str = "build/external-kits/external-kit-metadata.json";
 
 #[derive(Deserialize, Debug, Clone, PartialEq)]
+#[serde(rename_all = "kebab-case")]
```
Comment (webern) [inline tools/buildsys-config/src/lib.rs:9]: I don't understand why this changed.

(seems more like camelCase, but I guess that doesn't matter)

File: tools/buildsys/src/manifest.rs [hunk 14]
```diff
@@ -322,11 +324,44 @@ impl Manifest {
         Ok(kits)
     }
 
+    /// List all external kits needed for the build in the format of "<vendor>-<kit_name>"
+    pub fn external_kit_dependencies<P>(&self, build_dir: P) -> Result<Vec<String>>
```
Comment (webern) [inline tools/buildsys/src/manifest.rs]: I don't think this function has anything to do with the manifest, does it? It's not using `&self` and seems unrelated to the existence of the manifest file. Likely there should be `::load(path)` function on `ExternalKitMetadataView` and this function should be a member of that object.

So maybe `ExternalKitMetadataView::load(path).list()`.

Anyway I don't think it belongs on `Manifest`

File: twoliter/embedded/build.Dockerfile (hunk 9 above)
Comment (webern) [reply twoliter/embedded/build.Dockerfile:158]: If that transformation is required, can it be done in the rust code that provides the list instead?

File: twoliter/embedded/build.Dockerfile (hunk 8 above)
Comment (webern) [reply twoliter/embedded/build.Dockerfile:162]: I think it was for debugging in the previous block. Possibly removing it was forgotten. It seems like it could come in handy. I could go either way on removing it.

**Comment by jmt-lab** [comment]
Updated to matt's comments

File: tools/buildsys-config/src/lib.rs (hunk 13 above)
Comment (jmt-lab) [reply tools/buildsys-config/src/lib.rs:9]: camelCase would do the same, when deserializing they need to be amd64, arm64 not Amd64, Arm64

**Review by webern — APPROVED** [review]
